### PR TITLE
Afflictus Emittus and more

### LINKED ISSUE #26649: Virology symptom suggestion: Afflictus Emittus
image

Probably stage 4
Makes you constantly have an emitter beam coming out of your face.
It does way less damage than a normal emitter but can still rarely break lockers and make shards radiate

### PR DESCRIPTION
**Pull request by DeityLink**
engi_beam
Emitter Gogglesimage
image
image
image

Fixes #26649

* Emitter beams are now pooled, which should improve their CPU usage by a wide margin, especially when toggling them on or off.
* Added code hooks for behaviours tied to player movement.
* Added a new experimental type of glasses, the Emitter Goggles. When putting them on, the player will emit a 6-tile long emitter beam with a new look, but that otherwise behaves just like a regular emitter, except it works without need for an external power source, and the player can move around wearing them. However the beam deals very little damage while the player moves, it reaches full power when standing still. The only ways to interrupt the beam are to either remove the goggles, fall down, or fall unconscious. The toggle button is a 5 dollar DLC, coming this summer. Oh and there's actually no way of obtaining it outside of admin interference. The plan for now is to cautiously bus it, see how the players handle it, potentially (definitely) nerf it, and then add legit ways to obtain it in-game.
* Added a new experimental Stage 4 disease effect, Afflictus Emittus, suggested by @kilozombie . Infected individual who activate this effect will start emitting beams akin to those of the Emitter Goggles. At Strength 1 though the beam only deals a third of a regular emitter's damage, but it can be mutated up to Strength 3 for the same damage as a regular emitter. Mice can be infected, but their beam deals very little damage, and has a reach of only 4 tiles. The same ways that interrupt the beam on the Emitter Goggles will stop the beam caused by Afflictus Emitus, however added to that, the symptom will stop working if the infected human have no eyes, highly damaged eyes, or robotic eyes.  However, it will also regenerate said eyes, even if they have been extracted (though in the case of robotic eyes, they will pop-off first). As with the goggles, there are no ways to actually obtain the effect outside of admin interference. We'll have to see how well the server handles it from both a gameplay and CPU standpoint, then we might add it among the possible effect mutations.

:cl:
* rscadd: When extracting eyes from a human, the iris on the eye object matches the color of the eye
* rscadd: Added a new admin-only (for now) experimental type of glasses, the Emitter Goggles.
* rscadd: Added a new admin-only (for now) experimental Stage 4 disease effect, Afflictus Emittus, suggested by kilozombie.

### CODE REVIEW

**Comment by ghost** [comment]
laser cheese when

**Comment by BomberBro** [comment]
I really like all the changes but it feels a bit unatomic to have it all rolled into one PR

**Comment by Irradiation** [comment]
Prepare for unforeseen consequences.

**Comment by DeityLink** [comment]
@BomberBro I agree, but that reflects how chaotic the day went. It went like this:

1. Ok let's try to code @kilozombie 's virus symptom idea
2. Ok that's done, now let's test it
3. Oh no the server lags when you've got 50 mobs infected with it, I need to workaround that
4. Ok let's try porting the effect to glasses instead, it shouldn't lag with a single mob, and also reduce the range from infinity (lol) to 6 tiles.
5. Darn it still lags a bit when I move, maybe I can try pooling the beams to fix it?
6. Oh cool! Pooling fixed all the lag issues...but now there's dozens other issues caused by the pooling, gotta reset those vars, etc
7. Ok finally the goggles work without any bugs, the emitters are still working properly, maybe I can save the virus symptom after all now that pooling reduces the lag
8. Nice, I can have dozens of infected mice and it doesn't lag at all anymore, now to do the finishing touches, so the symptom takes the eyes of human mobs into account

So basically stuff was done in a pretty chaotic order, if I knew from the start that I'd have to pool beams I might have made a separate PR, but the pooling code and movement hooks were highly tied with the goggles by then.

**Comment by kilozombie** [comment]
Holy shit

**Comment by Dacendeth** [comment]
Holy shit monkeyman emitter goggles shard engine yes. also does it look really weird while resting?

**Comment by Kurfursten** [comment]
You don't need to include in a changelog
* Code/backend things like new hooks/helpers/procs/vars, file cleanup, or performance improvements (unless you want players to be on the lookout for buggy behavior)
* Adminbus only things like the new goggles or symptom
* Lengthy explanations of specifics that are well suited for the wiki

**Comment by kilozombie** [comment]
The spritework is actually great by the way, even if there's not a ton of it. The lasers blasting mobs is hilarious

**Comment by Kharkov72** [comment]
-unatomic uncle
+++incredibly based mice lasers

**Comment by SkowronX** [comment]
YOU CAN ALL STOP STARING NOW

**Comment by ShiftyRail** [comment]
I can't wait for someone to PR it as a mice virus. 20/20

File: code/game/objects/effects/beam.dm [hunk 1]
```diff
@@ -61,10 +69,92 @@
...
+/obj/effect/beam/emitter/eyes/emit(var/spawn_by, var/_range=-1,var/charged = FALSE)
+	..()
+	if (charged)
+		charge_up()
+	else
+		var/turf/T = loc
+		var/old_dir = dir
+		var/old_steps = steps
+		spawn(5)
+			if (src && loc && loc == T && dir == old_dir && steps == old_steps)
```
Comment (ShiftyRail) [inline code/game/objects/effects/beam.dm]: This is a bit of a weird check, some of it is redundant.
`src` will never be false, if you want to know if your thing hasn't been marked for deletion, check `gcDestroyed` instead. `loc == T` will implicitly fail if `loc == null`.

File: code/game/objects/effects/beam.dm [hunk 2]
```diff
@@ -150,8 +240,10 @@
 	if(!OB)
 		OB = src
 	src._re_emit = 0
-	qdel(src)
+	//returnToPool(src)
```
Comment (ShiftyRail) [inline code/game/objects/effects/beam.dm]: Remove dead commented code, it's a bit confusing.

File: code/game/objects/effects/beam.dm [hunk 3]
```diff
@@ -185,8 +283,12 @@
 		BM.disconnect(0)
 	BM.target=AM
 	if(istype(AM))
+		if (!AM.on_moved)
```
Comment (ShiftyRail) [inline code/game/objects/effects/beam.dm:288]: I'm not sure when this would be false. All `atom/movable` have this event initialised.

File: code/game/objects/effects/beam.dm [hunk 4]
```diff
@@ -208,15 +310,19 @@
...
-			qdel(child)
+			returnToPool(child)
 	children.len = 0
 
 /obj/effect/beam/proc/disconnect(var/re_emit=1)
 	var/obj/effect/beam/_master=get_master()
 	if(_master.target)
-		if(ismovable(_master.target) && _master.target.on_moved)
+		if(ismovable(_master.target))
+			if (!_master.target.on_moved)
```
Comment (ShiftyRail) [inline code/game/objects/effects/beam.dm:327]: Same as above.

File: code/game/objects/effects/beam.dm [hunk 5]
```diff
@@ -225,17 +331,21 @@
...
+		_master.update_icon()
 
 /obj/effect/beam/Crossed(atom/movable/AM as mob|obj)
 	beam_testing("Crossed by [AM]")
 	if(!master || !AM)
 		beam_testing(" returning (!AM || !master)")
 		return
 
-	if(istype(AM, /obj/effect/beam) || (!AM.density && !istype(AM, /obj/effect/blob)))
+	if(istype(AM, /obj/effect/beam) || (!AM.density && !istype(AM, /obj/effect/blob)))// || (istype(AM, /obj/machinery/door) && !AM.opacity)
```
Comment (ShiftyRail) [inline code/game/objects/effects/beam.dm]: Why is there a dead code comment here ?

File: code/game/objects/items/candle.dm [hunk 6]
```diff
@@ -74,11 +74,14 @@
 	return 0
 
 /obj/item/candle/Crossed(var/obj/Proj)
-	if(isbeam(Proj))
-		//Can also be an obj/effect, but they both have the damage var
+	if(istype(Proj,/obj/item/projectile/beam))
```
Comment (ShiftyRail) [inline code/game/objects/items/candle.dm]: Why did you change this helper ?

File: code/modules/clothing/glasses/glasses.dm [hunk 7]
```diff
@@ -543,3 +543,95 @@ var/list/science_goggles_wearers = list()
...
+	previous_dir = null
+	previous_loc = null
+	..()
+
+/obj/item/clothing/glasses/emitter/proc/enable(var/mob/living/mob)
+	if (istype(emitter))
+		emitter.callOnStartMove |= "\ref[src]"
+		emitter.callOnStartMove["\ref[src]"] = "update_emitter_start"
+		emitter.callOnEndMove |= "\ref[src]"
+		emitter.callOnEndMove["\ref[src]"] = "update_emitter_end"
```
Comment (ShiftyRail) [inline code/modules/clothing/glasses/glasses.dm]: `|=` is slow. I think you can condense this with just `L["\ref[src]"] = "update_emitter_start"`, since `"\ref[src]"` and `"update_emitter_start"` are just strings (doing this with atoms as indexes causes issue for some reasons, I learned it in the runechat PR).

File: code/modules/mob/mob.dm [hunk 8]
```diff
@@ -1639,8 +1639,11 @@ Use this proc preferably at the end of an equipment loadout
 		return 1
 	if(!canface())
 		return 0
+	if (dir!=EAST)
```
Comment (ShiftyRail) [inline code/modules/mob/mob.dm]: Why `EAST` ? Pls comment your code

File: code/modules/mob/mob.dm [hunk 9]
```diff
@@ -1653,8 +1656,11 @@ Use this proc preferably at the end of an equipment loadout
 		return 1
 	if(!canface())
 		return 0
+	if (dir!=WEST)
```
Comment (ShiftyRail) [inline code/modules/mob/mob.dm]: Same as above

File: code/modules/mob/mob.dm [hunk 10]
```diff
@@ -1697,6 +1709,30 @@ Use this proc preferably at the end of an equipment loadout
 			src.callOnFace -= .
 
 
+/mob/proc/StartMoving()
+	var/datum/listener
+	for(. in src.callOnStartMove)
```
Comment (ShiftyRail) [inline code/modules/mob/mob.dm]: This is a really confusing usage of `.` for very little gain... Use a normal variable here.

File: code/modules/mob/mob.dm [hunk 11]
```diff
@@ -1697,6 +1709,30 @@ Use this proc preferably at the end of an equipment loadout
...
+		listener = locate(.)
+		if(listener)
+			call(listener,src.callOnStartMove[.])(src)
+		else
+			src.callOnStartMove -= .
+
+/mob/proc/EndMoving()
+	var/datum/listener
+	for(. in src.callOnEndMove)
```
Comment (ShiftyRail) [inline code/modules/mob/mob.dm]: Same as above.

File: code/modules/mob/mob.dm [hunk 12]
```diff
@@ -1697,6 +1709,30 @@ Use this proc preferably at the end of an equipment loadout
 			src.callOnFace -= .
 
 
+/mob/proc/StartMoving()
```
Comment (ShiftyRail) [inline code/modules/mob/mob.dm:1676]: Please explain the reasons of these wrappers in a comment above so that future coders can understand why those procs exists at a glance.

**Review by ShiftyRail — COMMENTED** [review]
Please add comments to your code

**Comment by DeityLink** [comment]
@Dacendeth the beams stop firing when lying down.

@Kurfursten that wouldn't leave much to the changelog. Fair enough for the backend stuff but the point of the item and symptom is to get bused and observed so it can become non-adminbus-only, so I kinda need players who don't lurk the coder discord to know about it.

File: code/game/objects/effects/beam.dm (hunk 3 above)
Comment (DeityLink) [reply code/game/objects/effects/beam.dm:288]: That was to silence a runtime linked to beams trying to pass walls (turfs) as args. I'm not sure how that could even be possible due to the check for ismovable but you tell me.

File: code/game/objects/effects/beam.dm (hunk 4 above)
Comment (DeityLink) [reply code/game/objects/effects/beam.dm:327]: Same as above, those were two runtimes.

File: code/game/objects/items/candle.dm (hunk 6 above)
Comment (DeityLink) [reply code/game/objects/items/candle.dm]: Beams no longer have the damage var, gotta give projectiles a get_damage proc instead. I think this bit of code was risky anyway to assume that both types would always share that same var.

File: code/game/objects/effects/beam.dm (hunk 3 above)
Comment (ShiftyRail) [reply code/game/objects/effects/beam.dm:288]: http://www.byond.com/docs/ref/info.html#/proc/ismovable

BYOND proc not working as intended, it seems.

**Comment by SonixApache** [comment]
you know your :cl: is totally busted right

**Comment by DeityLink** [comment]
@SonixApache you missed the "edited by Kurfursten" up top.

File: code/modules/mob/mob.dm (hunk 12 above)
Comment (DeityLink) [reply code/modules/mob/mob.dm:1676]: added a comment the line before

**Comment by DeityLink** [comment]
Morning pass of fixes done.

**Comment by Blithering** [comment]
This is the best, something I've wanted for a while. Two requests though, 

1. can you have a pair of goggles with the opposite effect, blocking the beams cyclops style with a button to fire your pewpew laser vision

2. can you have the mouse beams fire sideways to represent poor binocular vision

**Comment by DeityLink** [comment]
@Blithering 
1. Possibly in a future PR, easily doable in theory.
2. >diagonal emitter beams

allow me to curl in the corner for a few minutes.

**Comment by ComradeDispenser** [comment]
Would be neat if like a welding mask or welding goggles being flipped down could block the eye beams, doesn't work for advanced welding goggles because auto-darkening lenses are weird.
